App Crashes When Tapping On Scan QR Code

#### General information

* **App version:** 0.9.0-beta1
* **App source:** GitHub
* **Android Version:** 11
* **Custom ROM:** LineageOS 18.1

#### Expected result
**What is expected?**
The app launches the QR Scanning normally. 

**What does happen instead?**
The app crashes just after tapping on "Scan QR Code".

#### Logcat
https://pastebin.com/7QLKWKab

#### Steps to reproduce
 
 * Launch andOTP.
 * Tap on the "+" button.
 * Tap on the "Scan QR Code" option.
 * Experiment the crash.

Can be reproduced on other phones with Android 11. Might be related to the following log lines.

```
FATAL EXCEPTION: main
Process: org.shadowice.flocke.andotp, PID: 11594
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.shadowice.flocke.andotp/org.shadowice.flocke.andotp.Activities.SecureCaptureActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'android.view.WindowInsetsController com.android.internal.policy.DecorView.getWindowInsetsController()' on a null object reference
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3431)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3595)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:85)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:135)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2066)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'android.view.WindowInsetsController com.android.internal.policy.DecorView.getWindowInsetsController()' on a null object reference
	at com.android.internal.policy.PhoneWindow.getInsetsController(PhoneWindow.java:3880)
	at org.shadowice.flocke.andotp.Activities.SecureCaptureActivity.initializeContent(SecureCaptureActivity.java:29)
	at com.journeyapps.barcodescanner.CaptureActivity.onCreate(CaptureActivity.java:21)
	at android.app.Activity.performCreate(Activity.java:8000)
	at android.app.Activity.performCreate(Activity.java:7984)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1309)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3404)
	... 11 more
```

--- a/app/src/main/java/org/shadowice/flocke/andotp/Activities/SecureCaptureActivity.java
+++ b/app/src/main/java/org/shadowice/flocke/andotp/Activities/SecureCaptureActivity.java
@@ -1,3 +1,25 @@
+/*
+ * Copyright (C) 2019-2021 Jakob Nixdorf
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
 package org.shadowice.flocke.andotp.Activities;
 
 import android.content.res.Configuration;

--- a/app/src/main/java/org/shadowice/flocke/andotp/Activities/SecureCaptureActivity.java
+++ b/app/src/main/java/org/shadowice/flocke/andotp/Activities/SecureCaptureActivity.java
@@ -25,9 +25,11 @@ package org.shadowice.flocke.andotp.Activities;
 import android.content.res.Configuration;
 import android.content.res.Resources;
 import android.os.Build;
+import android.view.Window;
 import android.view.WindowInsets;
 import android.view.WindowInsetsController;
 import android.view.WindowManager;
+
 import com.journeyapps.barcodescanner.CaptureActivity;
 import com.journeyapps.barcodescanner.DecoratedBarcodeView;
 
@@ -36,6 +38,25 @@ import org.shadowice.flocke.andotp.Utilities.Settings;
 import java.util.Locale;
 
 public class SecureCaptureActivity extends CaptureActivity {
+
+    /**
+     * Overwrites {@link CaptureActivity#initializeContent()} to:
+     * <ul>
+     *     <li>preventing the window from appearing in screenshots or from being viewed on
+     *     non-secure displays, if this was enabled in the app settings.</li>
+     *     <li>request hardware acceleration to be turned on.</li>
+     *     <li>hide all screen decorations like navigation and status bars.</li>
+     * </ul>
+     *
+     * <em>
+     *     Note:
+     *     {@link android.app.Activity#setContentView(int)} in {@link CaptureActivity#initializeContent()}
+     *     of super class needs to be called before {@link Window#getInsetsController()}, otherwise NPE will be thrown
+     *     because the top-level view of the current {@link Window}, containing the window decor, is not yet initialized.
+     * </em>
+     *
+     * @return the DecoratedBarcodeView
+     */
     @Override
     @SuppressWarnings("deprecation")
     protected DecoratedBarcodeView initializeContent() {
@@ -44,8 +65,19 @@ public class SecureCaptureActivity extends CaptureActivity {
         setTheme(settings.getTheme());
         setLocale(settings);
 
-        if (!settings.getScreenshotsEnabled())
+        //  This flag must be set before setting the content view of the activity or dialog.
+        getWindow().setFlags(
+                WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED,
+                WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED
+        );
+
+        // All window flags, that must be set before the window decoration is created, are already set
+        // so we are safe to call super here.
+        DecoratedBarcodeView barcodeScannerView = super.initializeContent();
+
+        if (!settings.getScreenshotsEnabled()) {
             getWindow().setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE);
+        }
 
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
             final WindowInsetsController insetsController = getWindow().getInsetsController();
@@ -54,12 +86,13 @@ public class SecureCaptureActivity extends CaptureActivity {
                 insetsController.setSystemBarsBehavior(WindowInsetsController.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE);
             }
         } else {
-            getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
+            getWindow().setFlags(
+                    WindowManager.LayoutParams.FLAG_FULLSCREEN,
+                    WindowManager.LayoutParams.FLAG_FULLSCREEN
+            );
         }
 
-        getWindow().setFlags(WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED, WindowManager.LayoutParams.FLAG_HARDWARE_ACCELERATED);
-
-        return super.initializeContent();
+        return barcodeScannerView;
     }
 
     private void setLocale(Settings settings) {

